### DISCUSSION TITLE
Custom JavaFX webview plugin [help]

### DISCUSSION THREAD

**Discussion by vickykr26941** [nroduit/Weasis#595] [Q&A]
Hi Devs,

I am new to Weasis and OSGI framework. i am trying to build a custom dockable plugin to integrate a simple WebView for reporting purposes. I have added the necessary JavaFX dependencies (e.g., javafx-base, javafx-web, javafx-swing, etc.) and included all the imports in the base.json file under extra imports. However, I am encountering a ClassNotFoundError at runtime when trying to call any methods of  classes that contain JavaFX-related imports.

Error logs:
30.09.2024 20:28:07.172 *INFO* [AWT-EventQueue-0] org.weasis.dicom.explorer.LoadDicom: End of loading DICOM locally
Exception in thread "AWT-EventQueue-0" java.lang.NoClassDefFoundError: javafx/beans/value/ObservableValue
	at org.weasis.dicom.viewer2d.dockable.DisplayTool.lambda$iniTree$3(DisplayTool.java:159)
	...
Caused by: java.lang.ClassNotFoundException: javafx.beans.value.ObservableValue
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	...
Here’s what I added to the base.json under the extra dependencies section:

base.json extra plugins.
{
  "code": "org.osgi.framework.system.packages.extra",
  "value": "${framework.system.packages.extra.basic}, javafx.animation, javafx.beans, javafx.beans.value, javafx.scene, javafx.embed.swing, javafx.scene.web, javafx.beans, javafx.beans.value ...",
  "description": "List of additional packages for OSGI framework (by default only java.* is visible in main class loader)",
  "type": "A",
  "category": "FELIX_CONFIG"
}
Has anyone encountered a similar issue or knows how to resolve this ClassNotFoundError or there any way i can integrate a webview plugin. ? Any help would be appreciated.

**Comment by vickykr26941** [comment]
@nroduit 
Or is there any way can i integrate the a simple webview plugin in installed versions ? i tried with this documentation (https://weasis.org/en/basics/customize/build-plugins/) 
Created a sample plugin with weasis-plugin-dicom-viewer-archetype . but may be i am missing something. sample plugin is not visible even in older versions i tried this.  Is there a way i can integrate the simple webview plugin using javaFX ?

**Comment by nroduit** [comment]
JavaFX was evaluated but not chosen several years ago, as there were significant performance problems on canvas for 2D rendering on images.

Using JavaFX properly requires quite a bit of transformation in the launcher and OSGI configuration.

Unfortunately, I can only advise you not to use JavaFX.

**Reply by vickykr26941** [reply]
Thank you for the response.
I am trying to the same without JavaFX now.

I am currently trying to embed a reporting panel into Weasis, where the goal is to load a browser-based editor that allows radiologists and doctors to create reports and download them as PDFs. Additionally, the embedded browser could be used for other related functionalities as a plugin.

I’ve already created a sample plugin for this purpose:
![image](https://github.com/user-attachments/assets/c27a7162-0931-4f15-a255-72fff0d98b42)


In an independent project, I successfully embedded a browser using the third-party library JCEF (Java Chromium Embedded Framework). However, I’m facing an issue integrating this into Weasis because the Maven plugin I’m using is not OSGi-ready.

Here is the Maven dependency I’m using (with transitive dependencies):
```
<dependency>
    <groupId>me.friwi</groupId>
    <artifactId>jcefmaven</artifactId>
    <version>126.2.0</version>
</dependency>
```

Do you have any suggestions on how to make this Maven dependency OSGi-compatible so it can be used in the project? Any help would be greatly appreciated!

**Comment by nroduit** [comment]
I've had a quick look at the project and I think that unfortunately it hasn't been designed to handle different classloaders as in OSGI. In fact, OSGI already does this job more cleanly and in Weasis you can see that packaging is [relatively simple for JOGL](https://github.com/nroduit/Weasis/tree/master/weasis-dicom/weasis-dicom-3d) (also used in jcef). You just need to find the right configuration for the main bundle with the import/export packages, and the native packages are bundle fragments that load automatically by OSGI.

Perhaps a simpler alternative for managing an independent third-party application is to use [launchers](https://weasis.org/en/tutorials/launcher-external/) to transmit patient/exam information to a third-party application.
